src/components/LibraryItem.tsx

```diff
@@ -54,7 +57,21 @@ export class LibraryItem extends React.Component<LibraryItemProps, LibraryItemSt
 
     constructor(props: LibraryItemProps) {
         super(props);
-        this.state = { expanded: false }; // Assign initial state
+
+        // Assign initial state
+        this.state = {
+            expanded: this.props.data.expanded,
+            visible: this.props.data.visible
+        };
+    }
+
+    componentWillReceiveProps(nextProps: LibraryItemProps) {
+        if (nextProps.data.visible !== this.state.visible) {
+            this.setState({ visible: nextProps.data.visible });
+        }
+        if (nextProps.data.expanded !== this.state.expanded) {
+            this.setState({ expanded: nextProps.data.expanded });
+        }
```

Comment:
Can you describe why is this `componentWillReceiveProps` needed (and what's the difference between this and the `constructor` above)? If this is needed, can we not merge both the `setState` into one? That way React.js doesn't have to queue up two state changes since they are aiming at the same component.